xdsl/builder.py

```diff
@@ -0,0 +1,131 @@
+from __future__ import annotations
+
+from typing import Callable, TypeAlias, overload
+
+from dataclasses import dataclass
+
+from xdsl.ir import OperationInvT, Attribute, Region, Block, BlockArgument
+from xdsl.dialects.builtin import FunctionType
+
+
+@dataclass
+class OpBuilder:
+    """
+    A helper class to construct IRs, by keeping track of where to insert an operation.
+    Currently can only append an operation to a given block, in the future will mirror the
+    API of `OpBuilder` in MLIR.
+
+    https://mlir.llvm.org/doxygen/classmlir_1_1OpBuilder.html
+    """
+
+    block: Block
+    """
+    Operations will be inserted in this block.
+    """
+
+    def insert(self, op: OperationInvT) -> OperationInvT:
+        """
+        Inserts `op` in `self.block` at the current insertion point.
+        """
+
+        self.block.add_op(op)
+        return op
+
+    @staticmethod
+    def region(func: Callable[[OpBuilder], None]) -> Region:
+        """
+        Generates a region given a function.
+        """
+
+        block = Block()
+        builder = OpBuilder(block)
+
+        func(builder)
+
+        return Region.from_block_list([block])
+
+    @staticmethod
+    def _callable_region_args(
+        types: tuple[list[Attribute], list[Attribute]]
+    ) -> Callable[[_CallableRegionFuncType], tuple[Region, FunctionType]]:
+        """
+        Constructs a tuple of (Region, FunctionType) for a region that takes some
+        arguments, and may return some results. The types of the arguments and results
+        are passed in the `types` parameter.
+        """
+
+        input_types, return_types = types
+
+        def wrapper(
+                func: _CallableRegionFuncType) -> tuple[Region, FunctionType]:
+
+            block = Block.from_arg_types(input_types)
+            builder = OpBuilder(block)
+
+            func(builder, block.args)
+
+            region = Region.from_block_list([block])
+            ftype = FunctionType.from_lists(input_types, return_types)
+            return region, ftype
+
+        return wrapper
+
+    @staticmethod
+    def _callable_region_no_args(
+            func: Callable[[OpBuilder], None]) -> tuple[Region, FunctionType]:
+        """
+        Constructs a tuple of (Region, FunctionType) for a region that takes no arguments
+        and returns no results.
+        """
+
+        @OpBuilder._callable_region_args(([], []))
+        def res(builder: OpBuilder, args: tuple[BlockArgument, ...]) -> None:
+            func(builder)
+
+        return res
+
+    @overload
+    @staticmethod
+    def callable_region(
```

Comment:
Same here, we should probably remove the return types.
Also, why do we have a tuple of lists, instead of two arguments anyway?